rustls/src/client/tls12.rs

```diff
@@ -111,7 +111,7 @@ mod server_hello {
 
             // See if we're successfully resuming.
             if let Some(ref resuming) = self.resuming_session {
-                if resuming.session_id == self.session_id {
+                if !resuming.session_id.is_empty() && resuming.session_id == self.session_id {
```

Comment:
I also think this all is hard to understand because there are implicit assumptions here that aren't documented and not documented in the type system.

Here we have `self.resuming_session` as a TLS 1.3 session where we don't store the session ID, so it is empty, but in this TLS 1.2 code we hadn't accounted for that possibility. Ideally, a TLS 1.3 session shouldn't even have a `session_id` that is comparable to a TLS 1.2 session_id. That is, this line above should have been a compile-time error. So I think a better solution would be to change the type of `resuming.session_id` to `Option<SessionID>` or even clearer define a separate type:
```
enum ResumptionSessionID {
    /// We don't store a session ID for TLS 1.3 sessions.
    Tls13,

    /// A non-empty session ID is always stored for TLS 1.2 sessions.
    ///
    /// The value of the session ID is always non-empty.
    Tls12(SessionID)
}
```

Reply:
Yes, I spent quite a bit of time yesterday trying to understand what was going on here and there are a number of things that could be improved. Unfortunately yesterday was my last day of work before my two-week vacation so I'm not sure how much time I'll be able to spend on this in the next week or so.

The improvement you suggested sounds good, but I think maybe we should go further? Some things I've come up with so far:

* We could start encoding the protocol version into the session cache key, making it an opaque `&[u8]`.
* We could change the type representation of the `SessionTicket` extension to use a nested enum rather than separate variants, where the nested enum's variants could be more explicit about the representative version.
* We should clarify what the intention is in preparing a `ClientHello` in terms of sending a 1.3 `PreSharedKeyOffer`, a 1.2 `SessionTicket` or both.
* In the client's 1.2 `ExpectFinished` state, it looks dubious to me that we can save the session if either the session ID or the ticket is empty (we only return early if both are).
* It looks like the client's 1.2 code might save an 1.2 session based on a session ID and a (presumably empty?) ticket from what was actually a cached 1.3 session.
* We might move any version-dependent fields from `ClientSessionValue`/`ServerSessionValue` into a version-variant enum, to make it more obvious what's going on there.

Yesterday I was still looking into the second problem @tanyav2 mentioned in #797, about which extensions the client sends in a 1.2-after-cached-1.3 scenario, so I just ran out of time. I dug through the RFCs to better understand the interactions but all I could find is that 1.3 obsoletes 5077, which doesn't quite make it clear what we should do here.

Nonetheless, I hope this is still a usable fix and a test case for the most egregious problem.